### PR TITLE
refactor: Introduce the new `FontBook`

### PR DESCRIPTION
**Pull request by agisilaos**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR I introduce the new `FontBook` so eventually we can get rid of `FontScheme` and the old way of handling Fonts. In addition to that, I renamed the old `DynamicFontButton` and `DynamicFontLabel` to `LegacyDynamicFontButton` & `LegacyDynamicFontLabel`

Last but not least I created two new classes `DynamicFontButton` and `DynamicFontLabel` with a brand new initializer using the new `FontStyle enum` and `static func font(for style: FontStyle) - > UIFont`. 

Deprecation messages/warnings have been added to both `LegacyDynamicFontButton` & `LegacyDynamicFontLabel`.

More information about the font changes can be found in this [confluence document](https://wearezeta.atlassian.net/wiki/spaces/IOS/pages/840794215/Fonts+on+Wire+iOS). 

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:1]: Is a new class really necessary? Another way would be to add a new initializer accepting the `FontStyle` and marking the existing initializer accepting the `FontSpec` as deprecated.

Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:1]: Done! PR is ready for another round of review.

File: wire-ios/WireCommonComponents/FontBook.swift
Comment (johnxnguyen) [inline wire-ios/WireCommonComponents/FontBook.swift:1]: What do you think about snapshot testing each font and content size category?

Comment (agisilaos) [reply wire-ios/WireCommonComponents/FontBook.swift:1]: This has been already taken care of.

File: wire-ios/Wire-iOS/Sources/Helpers/DynamicFontButton.swift [hunk 1]
```diff
@@ -24,11 +24,20 @@ import WireCommonComponents
...
+    init(style: UIFont.FontStyle = .body) {
+        // Not needed when we use a font style.
+        onRedrawFont = { return nil }
+        super.init(frame: .zero)
+        self.titleLabel?.font = .font(for: style)
+        titleLabel?.adjustsFontForContentSizeCategory = true
+    }
+
+    @available(*, deprecated, message: "Use `init(style:)` instead")
     init(fontSpec: FontSpec = .normalRegularFont) {
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/Helpers/DynamicFontButton.swift:39]: Should we already create a ticket to remove this deprecated initialize?

File: wire-ios/Wire-iOS/Sources/Helpers/DynamicFontButton.swift [hunk 2]
```diff
@@ -24,11 +24,20 @@ import WireCommonComponents
...
-        self.fontSpec = fontSpec
+        self.onRedrawFont = { return fontSpec.font }
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/Helpers/DynamicFontButton.swift]: I think `return` can be omitted:
```suggestion
        self.onRedrawFont = { fontSpec.font }
```

File: wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift [hunk 3]
```diff
@@ -21,31 +21,52 @@ import WireCommonComponents
...
+        onRedrawFont = { return nil }
+        super.init(frame: .zero)
+        self.text = text
+        self.textColor = color
+        self.font = .font(for: style)
+        self.adjustsFontForContentSizeCategory = true
+    }
+
+    @available(*, deprecated, message: "Use `init(text:style:color)` instead")
+    init(
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:48]: Same question:
> Should we already create a ticket to remove this deprecated initialize?

**Review by markusfassbender — APPROVED** [review]
Great improvement! Thanks 🤘

File: wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:48]: Yes, I will do that.

File: wire-ios/Wire-iOS/Sources/Helpers/DynamicFontButton.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/Helpers/DynamicFontButton.swift:39]: Yes! I will create the ticket! Thank you for the reminder!

**Review by David-Henner — APPROVED** [review]

**Pull request merged**